tests/utils/tensorflow/test_numpy.py

```diff
@@ -0,0 +1,37 @@
+import pytest
+import tensorflow as tf
+import numpy as np
+import rasa.utils.tensorflow.numpy
+import json
+from typing import Optional, Dict, Any
+
+
+class NumpyEncoder(json.JSONEncoder):
+    def default(self, obj):
+        if isinstance(obj, np.ndarray):
+            return obj.tolist()
+        return json.JSONEncoder.default(self, obj)
+
+
+@pytest.mark.parametrize(
+    "value, expected_result",
+    [
+        ({}, {}),
+        ({"a": 1}, {"a": 1}),
+        ({"a": tf.zeros((2, 3))}, {"a": np.zeros((2, 3))}),
+    ],
+)
+def test_values_to_numpy(
+    value: Optional[Dict[Any, Any]], expected_result: Optional[Dict[Any, Any]]
+):
+    actual_result = rasa.utils.tensorflow.numpy.values_to_numpy(value)
+    actual_result_value_types = [
+        type(value) for value in sorted(actual_result.values())
+    ]
+    expected_result_value_types = [
+        type(value) for value in sorted(actual_result.values())
+    ]
+    assert actual_result_value_types == expected_result_value_types
+    assert json.dumps(actual_result, sort_keys=True, cls=NumpyEncoder) == json.dumps(
```

Comment:
why do we need the `json.dumps` test? Can't we compare `expected_result == actual_result`?

Reply:
the `class NumpyEncoder` makes it a very weird way to test the correct order. I also don't think the order should change. Last option would to call `sort` on the results instead of mis-using the `json` library.

Reply:
Dictionaries have no default comparison for numpy arrays, so `expected_result == actual_result` doesn't work. Got this from [here](https://stackoverflow.com/a/47626762).

Reply:
So we'd have to write a function that compares the keys and values of the dict, and before it compares the dict values it has to check the type of the object there, and if it is a numpy array then it has to use the `tolist` method or the numpy comparison function, and all this should happen recursively down into list and dict values of the dict, which is a mess and easily solved with the json library.

Reply:
And I thought order is not guaranteed in dicts?

Reply:
> Dictionaries have no default comparison for numpy arrays

It should be the numpy arrays which provide the `__eq__` implementation 🤔  
How about using `numpy.testing.assert_equal(actual_result, expected_result)`?

> And I thought order is not guaranteed in dicts?

Order in Dicts is guaranteed since 3.6 (strictly: 3.7): https://softwaremaniacs.org/blog/2020/02/05/dicts-ordered/

Reply:
If you do `actual_result == expected_result` you get `ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()`. I've changed it to use `assert_equal` now, but I find it less elegant and harder to read.